# **Youtube Comments Scrapping**

List Video Youtube:

1) **Keterangan Pers Menkomdigi Terkait Transformasi Digital dan E-Government, Jakarta, 13 Januari 2025**

Channel: Sekretariat Presiden

Link: https://www.youtube.com/watch?v=1BmayjFMKGc

Total Komentar: 88

2) **Explained: Industri 4.0 | Transformasi Digital dan TelkomMetra**

Channel: TelkomMetra Official

Link: https://www.youtube.com/watch?v=iopb84dyQfk

Total Komentar: 30

   
3) **Salah Kaprah Transformasi Digital | Dharma S, Presiden Direktur Microsoft Indonesia [Part 2]**

Channel: Dr. Indrawan Nugroho

Link: https://www.youtube.com/watch?v=fneZo5_pox4

Total Komentar: 32


4) **Langkah Besar Transformasi Digital Indonesia**

Channel: Kementerian PANRB

Link: https://www.youtube.com/watch?v=8Kb3xfYXkZE

Total Komentar: 33


5) **Transformasi Digital, Siapkah Indonesia? - BERKAS KOMPAS**

Channel: KOMPASTV

Link: https://www.youtube.com/watch?v=71IoFn3Q2hs

Total Komentar: 34

### Scrapping (memakai YouTube Data API v3)

In [ ]:
from googleapiclient.discovery import build

# Inisialisasi API
api_key = "AIzaSyBdF-SlUeTMjDLye9p8NXl9sDRUF-qDbv8"
youtube = build('youtube', 'v3', developerKey=api_key)

# Fungsi untuk mengambil semua komentar dari sebuah video YouTube berdasarkan ID video
def get_all_comments(video_id):
    comments = []  # Inisialisasi daftar kosong untuk menyimpan komentar
    next_page_token = None  # Token untuk melacak halaman berikutnya

    # Loop untuk mengambil komentar hingga tidak ada lagi halaman
    while True:
        # Mengambil komentar dari API dengan parameter tambahan `maxResults`
        response = youtube.commentThreads().list(
            part="snippet",             # Menentukan bagian data yang akan diambil (snippet berisi informasi dasar komentar)
            videoId=video_id,           # ID video yang ingin diambil komentarnya
            textFormat="plainText",     # Format teks komentar diatur sebagai plain text (bukan HTML)
            maxResults=100,              # Maksimum komentar yang diambil dalam satu permintaan (100)
            pageToken=next_page_token    # Token untuk mengambil halaman berikutnya
        ).execute()

        # Menambahkan setiap komentar ke dalam daftar `comments`
        for item in response["items"]:
             # Mengakses teks komentar dari respons menggunakan struktur data JSON
            comment = item["snippet"]["topLevelComment"]["snippet"]["textDisplay"]
            # Menambahkan teks komentar ke dalam daftar `comments`
            comments.append(comment)

        # Melihat jumlah komentar yang terkumpul sejauh ini
        print(f"Total komentar sementara untuk video {video_id}: {len(comments)}")


        # Periksa apakah ada halaman berikutnya
        next_page_token = response.get("nextPageToken")
        if not next_page_token:  # Jika tidak ada `nextPageToken`, berarti sudah halaman terakhir
            break

    return comments  # Mengembalikan semua komentar yang dikumpulkan

# Panggil fungsi dengan beberapa video_id
video_ids = ["1BmayjFMKGc", "iopb84dyQfk", "fneZo5_pox4", "8Kb3xfYXkZE", "71IoFn3Q2hs"]
all_comments = []
for vid in video_ids:
    all_comments.extend(get_all_comments(vid))  # Mengambil semua komentar dari setiap video

Total komentar sementara untuk video 1BmayjFMKGc: 78
Total komentar sementara untuk video iopb84dyQfk: 22
Total komentar sementara untuk video fneZo5_pox4: 21
Total komentar sementara untuk video 8Kb3xfYXkZE: 33
Total komentar sementara untuk video 71IoFn3Q2hs: 33


Dapat dilihat diatas, bahwa tidak semua komentar dari video youtube terambil, hal ini bisa kemungkinan dikarenakan oleh 3 hal yaitu:

1. **Youtube API hanya scrapping komentar saja**: Youtube API hanya akan mengambil komentar dari user, bentuk komentar seperti reply komentar tidak diambil oleh API

2. **Komentar yang Dibatasi atau Dimoderasi**: YouTube API mungkin hanya mengizinkan akses ke komentar yang sudah terpublikasi dan tidak tersembunyi oleh pengaturan privasi atau moderasi, sehingga beberapa komentar tidak diambil


3. **Pembatasan API YouTube**: Beberapa video tidak mengizinkan pengambilan komentar dalam jumlah besar, atau API bisa saja membatasi jumlah komentar yang dapat diambil dalam satu waktu karena alasan keamanan atau privasi. Hal ini biasanya terjadi jika komentar dalam video tersebut diatur dengan pembatasan oleh pemilik video.

In [ ]:
# Buat dataframe
import pandas as pd
df = pd.DataFrame(all_comments, columns=["comment"])

# Tampilkan dataset di console sebelum disimpan
print(df)

# Simpan ke CSV
df.to_csv("youtube_comments.csv", index=False) # index=False artinya tidak ada index

                                               comment
0     https://youtu.be/5yT81Aw_1Jc?si=4IEpgFTEGmAx9RnX
1                                                    ❤
2    Judol serta penipuan masih marak di dunia digi...
3    Indonesia perlu kemandirian digital seperti bi...
4    JADI IBU ITU TIDAK HARUS BANYAK IKLAN DRAMA 🙏 ...
..                                                 ...
182  pemerintah memang telah berupaya menggenjot pe...
183  Menurut pendapat saya substansi yang di publik...
184  Apalagi nanti ada Jaringan 5G.\nTeknologi akan...
185  sehebat apapan teknologi digital, akan tetap k...
186             Honor febriasyah uda lunas nda iya....

[187 rows x 1 columns]


Dari output diatas, dapat kita ketahui bahwa terdapat 187 komentar yang terkumpul dari 5 video

# **Soal A**
[10%] Lakukan pre processing data dan text representation sebelum melakukan pemodelan
unsupervised learning, gunakan minimal 2 text representation approach.

**Membaca File**

In [ ]:
import pandas as pd

# Membaca file CSV yang berisi komentar YouTube
df = pd.read_csv("youtube_comments.csv")

# Menampilkan DataFrame untuk memeriksa data yang ada
print("Data awal sebelum preprocessing:")
df.head(10) # Menampilkan 10 baris pertama untuk pemeriksaan

Data awal sebelum preprocessing:


comment
0   https://youtu.be/5yT81Aw_1Jc?si=4IEpgFTEGmAx9RnX
1                                                  ❤
2  Judol serta penipuan masih marak di dunia digi...
3  Indonesia perlu kemandirian digital seperti bi...
4  JADI IBU ITU TIDAK HARUS BANYAK IKLAN DRAMA 🙏 ...
5  SITUS SITUS 🐶 ANJING KELAPARAN MASIH AMAN TERK...
6  KALAU SAYA JADI MENTRI KOMDIGI,,,!!! TIDAK AKA...
7  PAK PRABOWO JUJUR TEGAS AMANAH,,,,!!!TAPI PARA...
8  IBU MENTERI JANGAN HANYA BANYAK IKLAN DRAMA KO...
9  BUAT KOMDIGI,,, SAYA MAU LAPOR,,, MASIH BANYAK...

In [ ]:
df.tail(10)

comment
177  Menurut saya, transformasi digital menjadi kun...
178  Menurut pendapat saya. Setelah pandemi covid. ...
179  Transformasi digital sangat penting bagi perus...
180  Transformasi digital mengacu pada proses dan s...
181  menurut pendapat saya . usai covid ,dalam hal ...
182  pemerintah memang telah berupaya menggenjot pe...
183  Menurut pendapat saya substansi yang di publik...
184  Apalagi nanti ada Jaringan 5G.\nTeknologi akan...
185  sehebat apapan teknologi digital, akan tetap k...
186             Honor febriasyah uda lunas nda iya....

Dapat dilihat dari sini, dataset mengandung angka, tanda baca, emotikon, dan lainnya

## **Pre-processing**
**Langkah Preprocessing**
1. Cleansing: Menghapus tanda baca, angka, dan karakter
khusus lainnya.
2. Tokenization: Memecah kalimat menjadi kata-kata.
3. Remove Stopwords: Menghilangkan kata-kata umum yang tidak terlalu penting dalam analisis (kata seperti "dan", "yang").
4. Lemmatization: Mengubah setiap kata menjadi bentuk dasarnya untuk memudahkan analisis lebih lanjut.

### **1. Cleansing**
- Mengonversi semua teks ke huruf kecil
- Menghapus angka, tanda baca, dan karakter seperti !, •
- Menghapus emotikon
- Menghapus spasi ganda yang mungkin muncul akibat penghapusan karakter
- Menggunakan pustaka googletrans untuk menerjemahkan kata-kata yang bukan bahasa indonesia ke bahasa indonesia




In [ ]:
print(df.loc[33, 'comment'])

Pelayanan publik📱📲🎈🎈🎈🎈🎈🎈🌷🌷🌷🌷🌷🌻🌻🌻🌻🌻🌻🌻🌹🌹🌹🌹🦠🦠🐛🐛


Terdapat penggunakan emotikon

In [ ]:
print(df.loc[20, 'comment'])

Bismillah Semoga Indonesia Lebih Baik dan Maju
Berkat Pemimpin dan Pejabat Yang Tulus

Ya Allah Sehatkan dan Sukseskan Visi Misi
Bapak Prabowo dan Para Pejabat Pendukung 🤲🏻


Terdapat penggunaan spasi yang berlebih

In [ ]:
print(df.loc[29, 'comment'])

green turn red ... 🏃🏃🏃🏃🏃


Terdapat bahasa yang bukan bahasa indonesia

In [ ]:
# Install googletrans
!pip install langdetect deep-translator

In [ ]:
# Import libraries
import pandas as pd
import re
from langdetect import detect, LangDetectException
from deep_translator import GoogleTranslator

# Fungsi cleansing untuk teks
def clean_text(text):
    # Mengonversi ke huruf kecil
    text = text.lower()
    # Menghapus link (URL)
    text = re.sub(r'http\S+|www\S+', '', text)  # Menghapus URL yang dimulai dengan http, https, atau www
    # Menghapus emotikon dan simbol-simbol lain (seperti hati, checklist, dll.)
    text = re.sub(r'[^\w\s,]', '', text)  # Menghapus karakter selain huruf, angka, dan spasi
    # Menghapus emotikon (contoh: ❤️, 😍, ✅, dll.)
    text = re.sub(r'[^\x00-\x7F]+', '', text)  # Menghapus karakter Unicode non-ASCII (termasuk emotikon)
    # Menghapus karakter khusus dan spasi ganda
    text = re.sub(r'[^a-z\s]', '', text)
    text = re.sub(r'\s+', ' ', text).strip()  # Menghilangkan spasi berlebih

    # Mengecek apakah teks tidak kosong
    if text.strip():
        try:
            # Mendeteksi bahasa dan menerjemahkan jika bukan bahasa Indonesia
            detected_language = detect(text)
            if detected_language != 'id':
                text = GoogleTranslator(source=detected_language, target='id').translate(text)
        except LangDetectException:
            # Jika deteksi bahasa gagal, biarkan teks tetap tanpa perubahan
            pass
    else:
        # Jika teks kosong atau hanya spasi, biarkan kosong
        text = ''

    return text

# Menerapkan fungsi cleaning pada kolom 'comment'
df['cleaned_text'] = df['comment'].apply(clean_text)

# Menampilkan contoh hasil pembersihan
display(df[['comment', 'cleaned_text']])


comment  \
0     https://youtu.be/5yT81Aw_1Jc?si=4IEpgFTEGmAx9RnX   
1                                                    ❤   
2    Judol serta penipuan masih marak di dunia digi...   
3    Indonesia perlu kemandirian digital seperti bi...   
4    JADI IBU ITU TIDAK HARUS BANYAK IKLAN DRAMA 🙏 ...   
..                                                 ...   
182  pemerintah memang telah berupaya menggenjot pe...   
183  Menurut pendapat saya substansi yang di publik...   
184  Apalagi nanti ada Jaringan 5G.\nTeknologi akan...   
185  sehebat apapan teknologi digital, akan tetap k...   
186             Honor febriasyah uda lunas nda iya....   

                                          cleaned_text  
0                                                       
1                                                       
2    judol serta penipuan masih marak di dunia digi...  
3    indonesia perlu kemandirian digital seperti bi...  
4    jadi ibu itu tidak harus banyak iklan drama ta...  
..                                                 ...  
182  pemerintah memang telah berupaya menggenjot pe...  
183  menurut pendapat saya substansi yang di publik...  
184  apalagi nanti ada jaringan g teknologi akan le...  
185  sehebat apapan teknologi digital akan tetap ka...  
186           Honor Febriasyah adalah orang yang baik.  

[187 rows x 2 columns]

Dari output diatas, dapat dilihat bahwa teks sudah bersih

In [ ]:
# Data setelah cleaning
print(df.loc[20, 'cleaned_text'])

bismillah semoga indonesia lebih baik dan maju berkat pemimpin dan pejabat yang tulus ya allah sehatkan dan sukseskan visi misi bapak prabowo dan para pejabat pendukung


In [ ]:
# Data sebelum cleaning
print(df.loc[29, 'cleaned_text'])

hijau berubah menjadi merah


Dapat dilihat bahwa teks yang bukan bahasa indonesia, telah berhasil ditranslate ke bahasa indonesia.

### Split Data
Ratio:
1. Train: 80%
2. Test: 20%

In [ ]:
from sklearn.model_selection import train_test_split

# Split Data menjadi Training and Testing Sets
X_train, X_test = train_test_split(df['cleaned_text'], test_size = 0.2, random_state = 42)

In [ ]:
X_train.head()

95                        ok saya sudah dapat jawabannya
69     transformasi digital dan egovernment telah dit...
109    di birokrasi masih banyak ruang untuk transfor...
29                           hijau berubah menjadi merah
126                                            bismillah
Name: cleaned_text, dtype: object

In [ ]:
X_test.head()

185    sehebat apapan teknologi digital akan tetap ka...
78                                                      
55     judol masih merajarela di indonesia bu bersihk...
137    transformasi digital nasional melalui ina digi...
161    pendapat saya terkait isi video tersebut dapat...
Name: cleaned_text, dtype: object

### **2. Tokenization**
Memecah kalimat menjadi kata-kata



In [ ]:
# Libraries

# Install nltk
!pip install nltk

# Untuk tokenisasi
import nltk
from nltk.tokenize import word_tokenize
from nltk.util import ngrams

# Download resource NLTK untuk tokenisasi
nltk.download('punkt_tab')

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

#### Pengecekan Kata Teknis menggunakan bigram

In [ ]:
from collections import Counter

# Menggabungkan seluruh teks untuk tokenisasi
all_text = ' '.join(X_train)

# Tokenisasi seluruh teks
tokens = word_tokenize(all_text)

# Menyusun bigrams atau trigrams (pasangan kata atau tiga kata berturut-turut)
bigrams = ngrams(tokens, 2)  # Bigrams untuk pasangan kata
trigrams = ngrams(tokens, 3)  # Trigrams untuk tiga kata berturut-turut

# Menghitung frekuensi bigrams dan trigrams
bigram_freq = Counter(bigrams)
trigram_freq = Counter(trigrams)

# Menampilkan 10 bigrams dan trigrams paling umum
print("Most common bigrams:", bigram_freq.most_common(20))
print("Most common trigrams:", trigram_freq.most_common(20))

Most common bigrams: [(('transformasi', 'digital'), 43), (('di', 'indonesia'), 19), (('digital', 'dan'), 15), (('digital', 'yang'), 11), (('negeri', 'ini'), 11), (('teknologi', 'digital'), 11), (('institusi', 'kepolisian'), 10), (('digital', 'untuk'), 10), (('literasi', 'digital'), 10), (('ini', 'sudah'), 9), (('penegak', 'hukum'), 9), (('digital', 'di'), 8), (('bu', 'menteri'), 7), (('bangsa', 'ini'), 7), (('digital', 'ini'), 7), (('ekonomi', 'digital'), 7), (('yang', 'belum'), 6), (('hal', 'ini'), 6), (('terima', 'kasih'), 6), (('pendapat', 'saya'), 6)]
Most common trigrams: [(('transformasi', 'digital', 'dan'), 8), (('transformasi', 'digital', 'di'), 6), (('digital', 'di', 'indonesia'), 6), (('teknologi', 'digital', 'untuk'), 5), (('tantangan', 'transformasi', 'digital'), 5), (('transformasi', 'digital', 'yang'), 4), (('menurut', 'pendapat', 'saya'), 4), (('transformasi', 'digital', 'ini'), 4), (('institusi', 'kepolisian', 'ini'), 3), (('yg', 'melanggar', 'hukum'), 3), (('ini', 'sud

Berdasarkan output, terlihat tidak ada kata teknis

#### Tokenisasi

In [ ]:
# Fungsi untuk tokenisasi
def tokenize_text(text):
    return word_tokenize(text)

# Tokenisasi untuk data training dan testing
X_train_tokenized = X_train.apply(tokenize_text)
X_test_tokenized = X_test.apply(tokenize_text)

In [ ]:
# Data sebelum tokenisasi
X_train

95                        ok saya sudah dapat jawabannya
69     transformasi digital dan egovernment telah dit...
109    di birokrasi masih banyak ruang untuk transfor...
29                           hijau berubah menjadi merah
126                                            bismillah
                             ...                        
106    pandawa kabupaten bogor mengunakan microsoft form
14     komdigi cara hapus judolnya gmn ya ko masih bn...
92          ok udah dpt jawaban kuisnya ig nurlaelasarii
179    transformasi digital sangat penting bagi perus...
102    Cobalah untuk tidak terlalu banyak bergerak sa...
Name: cleaned_text, Length: 149, dtype: object

In [ ]:
# Data setelah tokenisasi
X_train_tokenized

95                  [ok, saya, sudah, dapat, jawabannya]
69     [transformasi, digital, dan, egovernment, tela...
109    [di, birokrasi, masih, banyak, ruang, untuk, t...
29                      [hijau, berubah, menjadi, merah]
126                                          [bismillah]
                             ...                        
106    [pandawa, kabupaten, bogor, mengunakan, micros...
14     [komdigi, cara, hapus, judolnya, gmn, ya, ko, ...
92     [ok, udah, dpt, jawaban, kuisnya, ig, nurlaela...
179    [transformasi, digital, sangat, penting, bagi,...
102    [Cobalah, untuk, tidak, terlalu, banyak, berge...
Name: cleaned_text, Length: 149, dtype: object

Dapat dilihat dari output diatas, kalimat sudah berubah jadi token

Namun, terdapat typo dan singkatan yang menggangu, sehingga perlu dilakukan typo handling juga

Contoh typo/singkatan:
1. dpt
2. gmn

#### **Tokenization + Typo Handling**

Untuk typo handling, beberapa cara sudah dicoba yaitu:
1. pyspellchecker --> Namun tidak support bahasa indonesia

2. hunspell --> bermasalah dalam install

3. idspell --> bermasalah dalam install

4. Textblob --> diterjemahkan ke bahasa inggris (co: tf idf menjadi if)

Cara typo handling yang digunakan:

--> membuat daftar kata bahasa Indonesia sendiri dan menggunakan pendekatan seperti pencarian kata terdekat atau distance metrics (Levenshtein Distance) untuk memperbaiki typo.

In [ ]:
from nltk.tokenize import word_tokenize
from difflib import get_close_matches

# Daftar kata bahasa Indonesia dan singkatan yang sering digunakan
indonesian_vocab = {'tidak', 'saya', 'yang', 'tetapi', 'siap', 'juga',
                    'ingin', 'untuk', 'bisa', 'bagaimana', 'terima kasih', 'izin', 'sudah'}

# Daftar khusus untuk singkatan dan penggantinya
abbreviation_map = {
    'tdk': 'tidak',
    'bs': 'bisa',
    'gmn': 'bagaimana',
    'tp': 'tapi',
    'utk': 'untuk',
    'jg': 'juga',
    'moga': 'semoga',
    'dpt': 'dapat',
    'bg': 'bagi',
    'tsb': 'tersebut',
    'yt': 'youtube'
}

# Fungsi sederhana untuk koreksi ejaan
def correct_spelling(word):
    # Periksa apakah kata ada dalam daftar singkatan
    if word in abbreviation_map:
        return abbreviation_map[word]

    # Jika bukan singkatan, cek typo dengan mencari kemiripan terdekat
    closest_match = get_close_matches(word, indonesian_vocab, n=1)
    return closest_match[0] if closest_match else word

# Fungsi untuk memproses teks menjadi token dan mengoreksi kesalahan
def tokenize_and_correct(text):
    # Ubah teks menjadi huruf kecil dan bagi menjadi kata-kata
    tokens = word_tokenize(text.lower())

    # Perbaiki setiap kata dalam daftar
    corrected_tokens = [correct_spelling(word) for word in tokens]

    return corrected_tokens

X_train_corrected = X_train.apply(tokenize_and_correct)
X_test_corrected = X_test.apply(tokenize_and_correct)

In [ ]:
X_train_tokenized

95                  [ok, saya, sudah, dapat, jawabannya]
69     [transformasi, digital, dan, egovernment, tela...
109    [di, birokrasi, masih, banyak, ruang, untuk, t...
29                      [hijau, berubah, menjadi, merah]
126                                          [bismillah]
                             ...                        
106    [pandawa, kabupaten, bogor, mengunakan, micros...
14     [komdigi, cara, hapus, judolnya, gmn, ya, ko, ...
92     [ok, udah, dpt, jawaban, kuisnya, ig, nurlaela...
179    [transformasi, digital, sangat, penting, bagi,...
102    [Cobalah, untuk, tidak, terlalu, banyak, berge...
Name: cleaned_text, Length: 149, dtype: object

In [ ]:
X_train_corrected

95                  [ok, saya, sudah, dapat, jawabannya]
69     [transformasi, digital, dan, egovernment, tela...
109    [di, birokrasi, masih, saya, yang, untuk, tran...
29                      [hijau, berubah, menjadi, merah]
126                                               [bisa]
                             ...                        
106    [pandawa, kabupaten, bogor, mengunakan, micros...
14     [komdigi, cara, hapus, judolnya, bagaimana, ya...
92     [ok, sudah, dapat, jawaban, kuisnya, ig, nurla...
179    [transformasi, digital, yang, penting, bagaima...
102    [cobalah, untuk, tidak, terlalu, saya, bergera...
Name: cleaned_text, Length: 149, dtype: object

Berdasarkan output diatas, dapat dilihat bahwa typo dan singkatan sudah di handle dengan baik

Contoh:
1. gmn --> bagaimana
2. dpt --> dapat

### **3. Remove Stopwords**
Menghilangkan kata-kata umum yang tidak terlalu penting dalam analisis



In [ ]:
# Libraries
import nltk
from nltk.corpus import stopwords

# Download stopwords dari NLTK jika belum ada
nltk.download('stopwords')
nltk.download('punkt')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [ ]:
# Daftar stopwords (gunakan bahasa Indonesia di sini)
stop_words = set(stopwords.words('indonesian'))

# Fungsi untuk menghapus stopwords setelah tokenisasi
def remove_stopwords(tokens):
    # Hapus stopwords
    filtered_tokens = [token for token in tokens if token.lower() not in stop_words]

    return filtered_tokens

# Penghapusan stopwords untuk data training dan testing
X_train_sw = X_train_corrected.apply(remove_stopwords)
X_test_sw = X_test_corrected.apply(remove_stopwords)


In [ ]:
# Data sebelum remove stopwords
X_train_corrected

95                  [ok, saya, sudah, dapat, jawabannya]
69     [transformasi, digital, dan, egovernment, tela...
109    [di, birokrasi, masih, saya, yang, untuk, tran...
29                      [hijau, berubah, menjadi, merah]
126                                               [bisa]
                             ...                        
106    [pandawa, kabupaten, bogor, mengunakan, micros...
14     [komdigi, cara, hapus, judolnya, bagaimana, ya...
92     [ok, sudah, dapat, jawaban, kuisnya, ig, nurla...
179    [transformasi, digital, yang, penting, bagaima...
102    [cobalah, untuk, tidak, terlalu, saya, bergera...
Name: cleaned_text, Length: 149, dtype: object

In [ ]:
# Data setelah remove stopwords
X_train_sw

95                                      [ok, jawabannya]
69     [transformasi, digital, egovernment, diterapka...
109    [birokrasi, transformasi, perpres, digital, pe...
29                               [hijau, berubah, merah]
126                                                   []
                             ...                        
106    [pandawa, kabupaten, bogor, mengunakan, micros...
14     [komdigi, hapus, judolnya, ko, bnyak, promosi,...
92                      [ok, kuisnya, ig, nurlaelasarii]
179    [transformasi, digital, perusahaan, beradaptas...
102                               [cobalah, bergerak, .]
Name: cleaned_text, Length: 149, dtype: object

Dari output, bisa dilihat kalau stopword dalam bahasa Indonesia seperti "saya" atau "sudah" telah terhapus

### **4. Lemmatization**
Mengubah setiap kata menjadi bentuk dasarnya untuk memudahkan analisis lebih lanjut.


In [ ]:
!pip install stanza
import stanza

# Mengunduh model bahasa Indonesia untuk Stanza
stanza.download('id')

# Inisialisasi pipeline Stanza dengan model bahasa Indonesia
nlp = stanza.Pipeline('id')

INFO:stanza:Downloaded file to /root/stanza_resources/resources.json
INFO:stanza:Downloading default packages for language: id (Indonesian) ...
INFO:stanza:File exists: /root/stanza_resources/id/default.zip
INFO:stanza:Finished downloading models and saved to /root/stanza_resources
INFO:stanza:Checking for updates to resources.json in case models have been updated.  Note: this behavior can be turned off with download_method=None or download_method=DownloadMethod.REUSE_RESOURCES


INFO:stanza:Downloaded file to /root/stanza_resources/resources.json
INFO:stanza:Loading these models for language: id (Indonesian):
| Processor    | Package      |
-------------------------------
| tokenize     | gsd          |
| mwt          | gsd          |
| pos          | gsd_charlm   |
| lemma        | gsd_nocharlm |
| constituency | icon_charlm  |
| depparse     | gsd_charlm   |

INFO:stanza:Using device: cpu
INFO:stanza:Loading: tokenize
INFO:stanza:Loading: mwt
INFO:stanza:Loading: pos
INFO:stanza:Loading: lemma
INFO:stanza:Loading: constituency
INFO:stanza:Loading: depparse
INFO:stanza:Done loading processors!


In [ ]:
# Fungsi untuk melakukan lemmatization menggunakan Stanza
def lemmatize_tokens(tokens):
    lemmatized_tokens = []

    for token in tokens:
        # Lakukan lemmatization dengan Stanza
        doc = nlp(token)  # Proses token melalui Stanza
        lemmatized_token = doc.sentences[0].words[0].lemma  # Ambil lemmatized token
        lemmatized_tokens.append(lemmatized_token)

    return lemmatized_tokens

# Terapkan lemmatizer
# df['lemmatized_comments_stanza'] = df['tokenized_comments_no_stopwords_1'].apply(lambda x: custom_lemmatizer(x, technical_terms))

# Aplikasikan lemmatization pada data token yang sudah di-split
X_train_lemmatized = X_train_sw.apply(lemmatize_tokens)
X_test_lemmatized = X_test_sw.apply(lemmatize_tokens)


In [ ]:
# Data sebelum lemmatization
X_train_sw

95                                      [ok, jawabannya]
69     [transformasi, digital, egovernment, diterapka...
109    [birokrasi, transformasi, perpres, digital, pe...
29                               [hijau, berubah, merah]
126                                                   []
                             ...                        
106    [pandawa, kabupaten, bogor, mengunakan, micros...
14     [komdigi, hapus, judolnya, ko, bnyak, promosi,...
92                      [ok, kuisnya, ig, nurlaelasarii]
179    [transformasi, digital, perusahaan, beradaptas...
102                               [cobalah, bergerak, .]
Name: cleaned_text, Length: 149, dtype: object

In [ ]:
# Data setelah lemmatization
X_train_lemmatized

95                                           [ok, jawab]
69     [transformasi, digital, egovernment, terap, in...
109    [birokrasi, transformasi, perpres, digital, ga...
29                                 [hijau, rubah, merah]
126                                                   []
                             ...                        
106    [pandawa, kabupaten, bogor, una, microsoft, form]
14     [komdigi, hapus, judol, ko, bnyak, promosi, ju...
92                         [ok, kuis, ig, nurlaelasarii]
179    [transformasi, digital, usaha, adaptasi, ubah,...
102                                  [cobalah, gerak, .]
Name: cleaned_text, Length: 149, dtype: object

Dari output diatas, proses lemmatization berjalan dengan baik

Contoh:
- Jawabannya >> jawab
- Diterapkan >> terap

## **Text Representation**

### TF-IDF

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Gabungkan kembali token menjadi string untuk TF-IDF
X_train_str = X_train_lemmatized.apply(lambda x: ' '.join(x))
X_test_str = X_test_lemmatized.apply(lambda x: ' '.join(x))

# Inisialisasi dan fit TF-IDF pada data training, lalu transform data training dan testing
tfidf_vectorizer = TfidfVectorizer()
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train_str)
X_test_tfidf = tfidf_vectorizer.transform(X_test_str)

print("TF-IDF training shape:", X_train_tfidf.shape)
print("TF-IDF testing shape:", X_test_tfidf.shape)

TF-IDF training shape: (149, 938)
TF-IDF testing shape: (38, 938)


Terdapat 938 unique tokens

In [ ]:
TFIDF_train=pd.DataFrame(X_train_tfidf.toarray(),columns=tfidf_vectorizer.get_feature_names_out())
TFIDF_train.head()

abdi  acakak  aceh  acu  adakarena  adaptasi  adil  adili  adopsi  adu  \
0   0.0     0.0   0.0  0.0        0.0       0.0   0.0    0.0     0.0  0.0   
1   0.0     0.0   0.0  0.0        0.0       0.0   0.0    0.0     0.0  0.0   
2   0.0     0.0   0.0  0.0        0.0       0.0   0.0    0.0     0.0  0.0   
3   0.0     0.0   0.0  0.0        0.0       0.0   0.0    0.0     0.0  0.0   
4   0.0     0.0   0.0  0.0        0.0       0.0   0.0    0.0     0.0  0.0   

   ...  wujud  yaa  yahoo  yarrobalamiiin  yeay  yono  youtube  yudikatif  \
0  ...    0.0  0.0    0.0             0.0   0.0   0.0      0.0        0.0   
1  ...    0.0  0.0    0.0             0.0   0.0   0.0      0.0        0.0   
2  ...    0.0  0.0    0.0             0.0   0.0   0.0      0.0        0.0   
3  ...    0.0  0.0    0.0             0.0   0.0   0.0      0.0        0.0   
4  ...    0.0  0.0    0.0             0.0   0.0   0.0      0.0        0.0   

   zalim  zina  
0    0.0   0.0  
1    0.0   0.0  
2    0.0   0.0  
3    0.0   0.0  
4    0.0   0.0  

[5 rows x 938 columns]

In [ ]:
TFIDF_test=pd.DataFrame(X_test_tfidf.toarray(),columns=tfidf_vectorizer.get_feature_names_out())
TFIDF_test.head()

abdi  acakak  aceh  acu  adakarena  adaptasi  adil  adili  adopsi  adu  \
0   0.0     0.0   0.0  0.0        0.0       0.0   0.0    0.0     0.0  0.0   
1   0.0     0.0   0.0  0.0        0.0       0.0   0.0    0.0     0.0  0.0   
2   0.0     0.0   0.0  0.0        0.0       0.0   0.0    0.0     0.0  0.0   
3   0.0     0.0   0.0  0.0        0.0       0.0   0.0    0.0     0.0  0.0   
4   0.0     0.0   0.0  0.0        0.0       0.0   0.0    0.0     0.0  0.0   

   ...    wujud  yaa  yahoo  yarrobalamiiin  yeay  yono  youtube  yudikatif  \
0  ...  0.00000  0.0    0.0             0.0   0.0   0.0      0.0        0.0   
1  ...  0.00000  0.0    0.0             0.0   0.0   0.0      0.0        0.0   
2  ...  0.00000  0.0    0.0             0.0   0.0   0.0      0.0        0.0   
3  ...  0.33199  0.0    0.0             0.0   0.0   0.0      0.0        0.0   
4  ...  0.00000  0.0    0.0             0.0   0.0   0.0      0.0        0.0   

   zalim  zina  
0    0.0   0.0  
1    0.0   0.0  
2    0.0   0.0  
3    0.0   0.0  
4    0.0   0.0  

[5 rows x 938 columns]

### Word2Vec

In [ ]:
# Library
!pip install gensim

from gensim.models import Word2Vec
import numpy as np

#### CBOW

In [ ]:
# Fungsi untuk menghasilkan vektor rata-rata dari setiap kalimat yang sudah ditokenisasi
def generate_word_vectors(tokenized_texts, word_vector_model, vector_length=50):
    X_vectors = []  # List untuk menyimpan vektor kata yang telah dihitung

    # Iterasi untuk setiap kalimat yang sudah ditokenisasi
    for tokens in tokenized_texts:
        # Ambil vektor kata yang valid untuk setiap token dalam kalimat
        # Token yang tidak ada dalam model Word2Vec akan diabaikan
        vectorized_tokens = [word_vector_model.wv.get_vector(word) for word in tokens if word in word_vector_model.wv]

        # Jika ada vektor kata yang valid, hitung vektor rata-rata
        if vectorized_tokens:
            # Menjumlahkan seluruh vektor kata yang valid dan membaginya dengan jumlah kata
            average_vector = sum(vectorized_tokens) / len(vectorized_tokens)
            X_vectors.append(average_vector)  # Menyimpan vektor rata-rata
        else:
            # Jika tidak ada vektor kata yang valid, simpan vektor nol dengan panjang sesuai vector_length
            X_vectors.append([0] * vector_length)

    # Mengembalikan list vektor yang dihasilkan
    return X_vectors


# Latih model Word2Vec dengan CBOW (sg=0)
w2v_model = Word2Vec(sentences=X_train_lemmatized, vector_size=50, window=5, min_count=3, sg=0)

# Menghasilkan vektor rata-rata untuk data training dan testing menggunakan fungsi generate_word_vectors
X_train_vectors_cbow = generate_word_vectors(X_train_lemmatized, w2v_model)
X_test_vectors_cbow = generate_word_vectors(X_test_lemmatized, w2v_model)

# Menampilkan dimensi hasil Word2Vec
print("Word2Vec training shape (CBOW):", len(X_train_vectors_cbow), len(X_train_vectors_cbow[0]) if X_train_vectors_cbow else 0)
print("Word2Vec testing shape (CBOW):", len(X_test_vectors_cbow), len(X_test_vectors_cbow[0]) if X_test_vectors_cbow else 0)

Word2Vec training shape (CBOW): 149 50
Word2Vec testing shape (CBOW): 38 50


In [ ]:
# Hasil output
X_train_vectors_cbow[0:2]

[[0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0],
 array([ 2.3270487e-03,  6.5611897e-04,  6.5193058e-04,  3.1768619e-03,
         3.1974740e-04, -4.0262495e-03,  7.7776276e-03,  5.6752926e-03,
        -7.5132200e-03, -2.4234296e-03, -1.4486533e-04, -8.9002149e-03,
         2.3671591e-03,  6.8485476e-03,  2.9761167e-04,  3.2359578e-03,
         6.4719468e-03,  1.5546381e-04, -7.6439106e-03, -8.0940882e-03,
         5.4464336e-03,  4.4424119e-04,  5.7057901e-03,  4.0200846e-03,
         8.7146703e-03,  2.0787995e-03, -5.8032656e-03,  7.6716812e-03,
        -4.2083180e-03,  2.0180799e-03,  3.4982287e-03,  4.5665526e-03,
         4.6554860e-03, -8.8443793e-03, -2.4791793e-03, -9.8477863e-04,
         1.1457788e-02,  5.3977757e-05,  3.2785088e-03, -2.9585811e-03,
         5.4203356e-03,  1.61

#### Skipgram

In [ ]:
from gensim.models import Word2Vec
import numpy as np

# Latih model Word2Vec menggunakan Skip-gram (sg=1)
w2v_model = Word2Vec(sentences=X_train_lemmatized, vector_size=50, window=5, min_count=3, sg=1)

# Menghasilkan vektor rata-rata untuk setiap kalimat
def generate_word_vectors(tokenized_texts, word_vector_model, vector_length=50):
    X_vectors = []  # List untuk menyimpan vektor kata yang telah dihitung

    # Iterasi untuk setiap kalimat yang sudah ditokenisasi
    for tokens in tokenized_texts:
        # Ambil vektor kata yang valid untuk setiap token dalam kalimat
        # Token yang tidak ada dalam model Word2Vec akan diabaikan
        vectorized_tokens = [word_vector_model.wv.get_vector(word) for word in tokens if word in word_vector_model.wv]

        # Jika ada vektor kata yang valid, hitung vektor rata-rata
        if vectorized_tokens:
            # Menjumlahkan seluruh vektor kata yang valid dan membaginya dengan jumlah kata
            average_vector = np.mean(vectorized_tokens, axis=0)  # menggunakan rata-rata
            X_vectors.append(average_vector)  # Menyimpan vektor rata-rata
        else:
            # Jika tidak ada vektor kata yang valid, simpan vektor nol dengan panjang sesuai vector_length
            X_vectors.append([0] * vector_length)

    return X_vectors

# Menghasilkan vektor rata-rata untuk data training dan testing
X_train_vectors_skipgram = generate_word_vectors(X_train_lemmatized, w2v_model)
X_test_vectors_skipgram = generate_word_vectors(X_test_lemmatized, w2v_model)

print("Word2Vec training shape:", len(X_train_vectors_skipgram), len(X_train_vectors_skipgram[0]) if X_train_vectors_skipgram else 0)
print("Word2Vec testing shape:", len(X_test_vectors_skipgram), len(X_test_vectors_skipgram[0]) if X_test_vectors_skipgram else 0)

Word2Vec training shape: 149 50
Word2Vec testing shape: 38 50


In [ ]:
# Hasil output
X_train_vectors_skipgram[0:2]

[[0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0],
 array([ 0.00580762,  0.00094069,  0.00510131, -0.05998148,  0.02058562,
        -0.07747319,  0.04566682,  0.10134516, -0.0592714 , -0.04588433,
        -0.0050062 , -0.07535463,  0.01498117,  0.03271525,  0.00881676,
         0.0390675 , -0.00542359, -0.00879286, -0.04755638, -0.04963858,
         0.01083193,  0.01074554,  0.06170864,  0.02151113,  0.04287024,
         0.00068634, -0.02595264,  0.01562552, -0.05915895,  0.00165022,
         0.00404161,  0.01880131, -0.03153094, -0.04726193, -0.02976515,
         0.00347652,  0.09691797, -0.03535385,  0.02472499, -0.05116149,
         0.0657088 ,  0.0088537 ,  0.03642222, -0.00383501,  0.13612665,
         0.00801254, -0.06337917, -0.02747851,  0.03300332, -0.01469716],
       dtype=float

# **Soal B** - Text Clustering
[15%] Lakukan pengelompokan komentar dengan menggunakan pendekatan clustering,
silahkan ujicoba beberapa nilai cluster (k) yang mungkin anda peroleh dari data tersebut dan
juga beberapa text repsesentation yang diperoleh dari 2a, berikan perbandingan silhoutte
score sebagai justifikasi jumlah cluster (k) yang akan anda pilih

In [ ]:
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
import matplotlib.pyplot as plt

In [ ]:
def evaluate_clustering_with_silhouette_score(X, min_k=2, max_k=10):
    """
    Fungsi ini melakukan clustering dengan K-Means dan menghitung Silhouette Score
    untuk berbagai jumlah cluster (k) dari min_k hingga max_k.

    Parameters:
    - X: Data input (vector representation seperti TF-IDF, BOW, Word2Vec, dll)
    - min_k: Nilai k minimum untuk clustering
    - max_k: Nilai k maksimum untuk clustering

    Returns:
    - silhouettes: Daftar Silhouette Score untuk setiap nilai k
    """
    silhouettes = []

    # Melakukan KMeans untuk berbagai nilai k
    for k in range(min_k, max_k + 1):
        kmeans_model = KMeans(n_clusters=k, random_state=42)
        cluster_labels = kmeans_model.fit_predict(X)

        # Hitung Silhouette Score untuk clustering tersebut
        silhouette_avg = silhouette_score(X, cluster_labels)
        silhouettes.append(silhouette_avg)

        # Menampilkan nilai Silhouette Score untuk k
        print(f"For n_clusters = {k}, The average silhouette score is: {silhouette_avg}")

    return silhouettes

In [ ]:
# Evaluasi clustering untuk TF-IDF
print("Evaluating TF-IDF:")
silhouettes_tfidf = evaluate_clustering_with_silhouette_score(X_train_tfidf, min_k=2, max_k=10)

Evaluating TF-IDF:
For n_clusters = 2, The average silhouette score is: 0.038779843710113275
For n_clusters = 3, The average silhouette score is: 0.016996161081227066
For n_clusters = 4, The average silhouette score is: 0.012139515650473706
For n_clusters = 5, The average silhouette score is: 0.01258353692787221
For n_clusters = 6, The average silhouette score is: 0.011577823128509598
For n_clusters = 7, The average silhouette score is: -0.0015065065007787406
For n_clusters = 8, The average silhouette score is: 0.0062033721540982354
For n_clusters = 9, The average silhouette score is: 0.0029394920884888826
For n_clusters = 10, The average silhouette score is: 0.010041575432971224


Dari atas bisa dilihat bahwa **TF-IDF** menghasilkan Silhouette Score yang sangat rendah, terutama setelah jumlah cluster lebih dari 2. Ini menunjukkan bahwa **TF-IDF** tidak mampu memisahkan cluster dengan baik, karena perbedaan antar cluster cenderung kecil. Dengan kata lain, representasi ini kurang efektif untuk clustering pada dataset ini.

In [ ]:
# Evaluasi clustering untuk Word2Vec CBOW
print("Evaluating Word2Vec CBOW:")
silhouettes_w2v_cbow = evaluate_clustering_with_silhouette_score(X_train_vectors_cbow, min_k=2, max_k=10)

Evaluating Word2Vec CBOW:
For n_clusters = 2, The average silhouette score is: 0.14531988820568326
For n_clusters = 3, The average silhouette score is: 0.1262500729104834
For n_clusters = 4, The average silhouette score is: 0.12268037708747963
For n_clusters = 5, The average silhouette score is: 0.13965343448509684
For n_clusters = 6, The average silhouette score is: 0.14057722731113306
For n_clusters = 7, The average silhouette score is: 0.1418741170925291
For n_clusters = 8, The average silhouette score is: 0.16040547565154037
For n_clusters = 9, The average silhouette score is: 0.16080800506389983
For n_clusters = 10, The average silhouette score is: 0.15604003329473462


Dari hasil diatas, terlihat bahwa **Word2Vec CBOW** menunjukkan Silhouette Score yang rendah, dengan nilai tertinggi sekitar **0.1608** pada **n_clusters = 9**. Nilai ini menurun lebih jauh seiring bertambahnya jumlah cluster.

 **Kesimpulan**: Word2Vec CBOW tidak memberikan pemisahan yang baik antara cluster, sehingga representasi ini kurang optimal untuk pengelompokan komentar.

In [ ]:
# Evaluasi clustering untuk Word2Vec Skipgram
print("Evaluating Word2Vec Skipgram:")
silhouettes_w2v_skipgram = evaluate_clustering_with_silhouette_score(X_train_vectors_skipgram, min_k=2, max_k=10)

Evaluating Word2Vec Skipgram:
For n_clusters = 2, The average silhouette score is: 0.5040452808882371
For n_clusters = 3, The average silhouette score is: 0.4955593868725949
For n_clusters = 4, The average silhouette score is: 0.4389057797684874
For n_clusters = 5, The average silhouette score is: 0.40076450613682146
For n_clusters = 6, The average silhouette score is: 0.39017283390276075
For n_clusters = 7, The average silhouette score is: 0.32598607555954723
For n_clusters = 8, The average silhouette score is: 0.3450504184902661
For n_clusters = 9, The average silhouette score is: 0.31630465018321857
For n_clusters = 10, The average silhouette score is: 0.2996502891163443


Dari hasil output diatas, bisa dilihat kalau **Word2Vec Skipgram** memberikan hasil yang cukup baik dengan **Silhouette Score tertinggi** pada **n_clusters = 2** (skor 0.51).

**Kesimpulan**: Meskipun Word2Vec Skipgram dapat memberikan hasil clustering yang lebih baik dibandingkan CBOW, hasil terbaik tetap didapatkan dari **BOW** pada **n_clusters = 2**.

### **Kesimpulan Analisa**:
Berdasarkan evaluasi hasil skor siluet dari ketiga text representation diatas, representasi **Skipgram** pada **n_clusters = 2** memberikan hasil terbaik karena Silhouette Score-nya yang tinggi, menunjukkan pemisahan cluster yang baik. Oleh karena itu, untuk clustering komentar, akan menggunakan **skipgram** sebagai representasi teks yang lebih efektif.

# **Soal C**
[5%] Lakukan persona analysis untuk cluster yang anda hasilkan, jelaskan informasi yang anda
peroleh mengenai cluster tersebut berdasarkan personanya.

## Clustering

In [ ]:
# Menjadikan X_train ke dalam bentuk DataFrame
X_train = X_train.to_frame()
print(X_train.columns)

X_test = X_test.to_frame()
print(X_test.columns)

Index(['cleaned_text'], dtype='object')
Index(['cleaned_text'], dtype='object')


In [ ]:
from sklearn.cluster import KMeans

# Clustering dengan K-Means (gunakan vektor rata-rata dari Word2Vec)
kmeans_w2v = KMeans(n_clusters=2, random_state=42)
kmeans_w2v.fit(X_train_vectors_skipgram)

# Prediksi cluster untuk data training dan testing
train_clusters = kmeans_w2v.predict(X_train_vectors_skipgram)
test_clusters = kmeans_w2v.predict(X_test_vectors_skipgram)

# Tambahkan hasil cluster ke DataFrame
X_train['k_means'] = train_clusters
X_test['k_means'] = test_clusters

# Menampilkan hasil untuk data train
print("Train Clustering Results:")
print(X_train[['cleaned_text', 'k_means']].head())

# Menampilkan hasil untuk data test
print("Test Clustering Results:")
print(X_test[['cleaned_text', 'k_means']].head())

Train Clustering Results:
                                          cleaned_text  k_means
95                      ok saya sudah dapat jawabannya        1
69   transformasi digital dan egovernment telah dit...        0
109  di birokrasi masih banyak ruang untuk transfor...        0
29                         hijau berubah menjadi merah        1
126                                          bismillah        1
Test Clustering Results:
                                          cleaned_text  k_means
185  sehebat apapan teknologi digital akan tetap ka...        0
78                                                            1
55   judol masih merajarela di indonesia bu bersihk...        0
137  transformasi digital nasional melalui ina digi...        0
161  pendapat saya terkait isi video tersebut dapat...        0


In [ ]:
# Fungsi untuk menghapus stopwords, dan pastikan hasilnya menjadi string
def remove_stopwords(text):
    stop_words = set(stopwords.words('indonesian'))  # Tentukan daftar stopwords
    # Tokenisasi dan hapus stopwords, kemudian gabungkan kembali menjadi string
    return " ".join([word for word in text.split() if word not in stop_words])

# Analisis kata-kata dalam setiap cluster dengan menghapus stopwords
for cluster_num in X_train['k_means'].unique():
    print(f"\nTop words for Cluster {cluster_num}:")
    cluster_data = X_train[X_train['k_means'] == cluster_num]['cleaned_text']

    # Hapus stopwords dari teks dalam cluster
    cluster_data_cleaned = cluster_data.apply(remove_stopwords)

    # Gabungkan semua teks dalam cluster dan pecah menjadi kata
    # Gunakan " ".join untuk menggabungkan list kata menjadi string
    words = " ".join(cluster_data_cleaned).split()

    # Hitung frekuensi kata
    word_freq = pd.Series(words).value_counts()

    # Tampilkan 10 kata teratas
    print(word_freq.head(10))



Top words for Cluster 1:
semoga     9
bu         9
sukses     8
judol      8
yg         7
ya         7
menteri    7
pandi      6
youtube    5
ig         5
Name: count, dtype: int64

Top words for Cluster 0:
digital         110
indonesia        71
transformasi     47
yg               32
masyarakat       32
teknologi        30
meningkatkan     28
pemerintah       23
hukum            23
negara           21
Name: count, dtype: int64


### **Analisis Cluster Berdasarkan Label**

#### **Cluster 0: Berkaitan dengan Digital Transformasi**
**Topik utama**:  
- **Teknologi**: Diskusi mengenai inovasi dan transformasi digital yang mempengaruhi berbagai sektor, termasuk pemerintahan dan masyarakat.
- **Indonesia**: Banyak pembicaraan tentang situasi sosial dan teknologi di Indonesia.
- **Pemerintah**: Fokus pada kebijakan yang mendukung kemajuan teknologi dan transformasi digital.
- **Masarakat**: Isu sosial yang melibatkan masyarakat luas dan dampak dari teknologi pada mereka.
- **Negara**: Berbicara tentang negara secara umum dalam konteks perkembangan teknologi dan kebijakan.

**Kata Kunci**:
- **Digital**: Berhubungan dengan digitalisasi dan teknologi.
- **Transformasi**: Berkaitan dengan perubahan besar dalam masyarakat atau industri.
- **Teknologi, Hukum, Negara**: Topik-topik yang berkaitan dengan regulasi dan implementasi teknologi.

**Deskripsi Singkat**:  
Cluster 0 berfokus pada topik-topik yang lebih berat dan terkait dengan transformasi digital, dampak sosial dari teknologi, serta isu-isu pemerintahan dan hukum di Indonesia. Ini menunjukkan ketertarikan pada perkembangan teknologi dan pengaruhnya terhadap kebijakan publik. Oleh karena itu, Cluster 0 mengandung topik yang **berkaitan dengan digital transformasi**

---

#### **Cluster 1: Tidak berkaitan dengan digital transformasi**
**Topik utama**:  
- **Harapan dan Sukses**: Berbicara tentang motivasi pribadi dan aspirasi untuk mencapai kesuksesan dalam kehidupan.
- **Media Sosial**: Pembahasan terkait media sosial seperti Instagram dan YouTube.
- **Motivasi**: Banyak kata-kata yang berhubungan dengan dorongan semangat dan harapan seperti "semoga sukses", "sukses", dll.

**Kata Kunci**:
- **Semoga, Sukses**: Kata yang sering digunakan untuk memberikan harapan atau doa untuk kesuksesan.
- **Mentri, Pandi**: Berhubungan dengan obrolan mengenai figur-figur populer.
- **YouTube, IG (Instagram)**: Media sosial yang populer di kalangan audiens muda.

**Deskripsi Singkat**:  
Cluster 1 berfokus pada topik yang lebih ringan dan pribadi, termasuk harapan, motivasi, dan diskusi di media sosial. Ini menunjukkan minat terhadap kehidupan sehari-hari, aspirasi pribadi, dan interaksi sosial yang terjadi di platform media sosial. Oleh karena itu, Cluster 1 mengandung topik **yang tidak berkaitan dengan digital transformasi**

---

### **Kesimpulan:**

- **Label 0**: Fokus pada **transformasi digital**
- **Label 1**: Tidak fokus pada **transformasi digital**



## Validasi menggunakan data test

In [ ]:
# Menampilkan hasil untuk data test
print("Test Clustering Results:")
print(X_test[['cleaned_text', 'k_means']].head())

Test Clustering Results:
                                          cleaned_text  k_means
185  sehebat apapan teknologi digital akan tetap ka...        0
78                                                            1
55   judol masih merajarela di indonesia bu bersihk...        0
137  transformasi digital nasional melalui ina digi...        0
161  pendapat saya terkait isi video tersebut dapat...        0


In [ ]:
# Menampilkan 5 contoh teks dari setiap cluster (label 0 dan label 1)
print("5 contoh teks dengan Label 0 (Cluster 0):")
print(X_test[X_test['k_means'] == 0]['cleaned_text'].head(5))  # Tampilkan 5 teks pertama dengan label 0
print("\n5 contoh teks dengan Label 1 (Cluster 1):")
print(X_test[X_test['k_means'] == 1]['cleaned_text'].head(5))  # Tampilkan 5 teks pertama dengan label 1

5 contoh teks dengan Label 0 (Cluster 0):
185    sehebat apapan teknologi digital akan tetap ka...
55     judol masih merajarela di indonesia bu bersihk...
137    transformasi digital nasional melalui ina digi...
161    pendapat saya terkait isi video tersebut dapat...
175    menurut saya transformasi digital akan memenga...
Name: cleaned_text, dtype: object

5 contoh teks dengan Label 1 (Cluster 1):
78                                                      
15     buaya ada an berita diterkam buaya diberbagai ...
75                      komen siap sumenep hadir ngo ami
154    ibu harus tuturnkan ilmu nanti ga ada berkahny...
122                                                     
Name: cleaned_text, dtype: object


Dapat dilihat dari output diatas bahwa Label 0 lebih berkaitan dengan pembicaraan transformasi digital dibadingkan dengan Label 1 yang sama sekali tidak berkaitan dengan topik transformasi digital

# **Soal D** - Topic Extraction
[15%] Anda tertarik untuk membandingkan hasil clustering dengan metode lain yaitu topic
extraction, sehingga anda perlu melakukan pemodelan topic extraction. Jelaskan metode topic
extraction yang anda lakukan

In [ ]:
import gensim
from gensim import corpora, models
from nltk.corpus import stopwords

# Fungsi build_dic
def build_dic(text):
    dictionary = corpora.Dictionary(text)
    dictionary.filter_extremes(no_below=1, no_above=0.2, keep_n=100000)
    return dictionary

    # PENJELASAN
    # no_below = 1 --> kata yang dihitung, frekuensi diatas 1
    # no_above = 0.2 --> kata kata yang sering muncul tidak boleh diatas 0.2
    # jadi kata yang terlalu sering dipake, itu perlu di drop
    # keep_n --> jumlah vocabulary yang disimpan

# Fungsi build_vec
def build_vec(text, dictionary):
    bow_corpus = [dictionary.doc2bow(doc) for doc in text]
    return bow_corpus

# Fungsi untuk menghitung TF-IDF
def vector_tfidf(bow_corpus):
    tfidf = models.TfidfModel(bow_corpus)
    corpus_tfidf = tfidf[bow_corpus]
    return corpus_tfidf

# Fungsi untuk melatih model LDA
def model_lda(dictionary, corpus_tfidf, num_topic, alpha, eta):
    lda_model = gensim.models.LdaMulticore(corpus_tfidf,
                                           num_topics=num_topic,
                                           id2word=dictionary,
                                           passes=10,
                                           workers=2,
                                           alpha=alpha,
                                           eta=eta)
    return lda_model

In [ ]:
# Membuat dictionary
dictionary = build_dic(X_train_lemmatized)

# Membuat BoW corpus dari X_train_lemmatized
bow_corpus = build_vec(X_train_lemmatized, dictionary)

# Model LDA (tanpa TF-IDF)
lda_model = model_lda(dictionary, bow_corpus, num_topic=3, alpha='symmetric', eta='auto')
                                              # Mau memakai 3 topik

# Melihat topik-topik yang ditemukan
topics = lda_model.print_topics(num_words=5)
for topic in topics:
    print(topic)

(0, '0.024*"judol" + 0.018*"semoga" + 0.013*"situs" + 0.013*"menteri" + 0.012*"bu"')
(1, '0.037*"transformasi" + 0.029*"tingkat" + 0.025*"masyarakat" + 0.022*"ekonomi" + 0.021*"terima"')
(2, '0.024*"hukum" + 0.014*"negara" + 0.013*"terima" + 0.013*"polisi" + 0.013*"narkoba"')


Dari output diatas, terlihat terdapat 3 topik yang terbentuk

In [ ]:
from gensim.models import CoherenceModel

# Fungsi untuk menghitung coherence score
def score_perf(lda_model, text, dictionary):
    coherence_model_lda = CoherenceModel(model=lda_model, texts=text, dictionary=dictionary, coherence='c_v')
    coherence_lda = coherence_model_lda.get_coherence()  # Mengambil nilai coherence
    return coherence_lda

# Menghitung coherence score menggunakan model LDA yang sudah ada
coherence_score = score_perf(lda_model, X_train_lemmatized, dictionary)

# Tampilkan coherence score
print("Coherence Score: ", coherence_score)

Coherence Score:  0.6204226544517031


Dari hasil diatas, dapat dilihat bahwa hasil coherence score diatas 0.3 yang berarti model LDA menghasilkan topik-topik yang cukup koheren dan kata-kata dalam setiap topik cenderung muncul bersama dalam dokumen. Skor ini menunjukkan bahwa **topik yang ditemukan cukup bermakna dan dapat dipahami dengan baik.**

# **Soal E**
[5%] Lakukan persona analysis kembali untuk topik yang anda peroleh, bandingkan apa yang
anda peroleh dari pendekatan clustering dan topic extraction. Paparkan hal-hal menarik yang
berhasil anda ekstrak.



## **Persona Analysis - Topic Prediction**

In [ ]:
# Kata-kata yang berperan untuk membangun topik

for idx, topic in lda_model.print_topics(-1):
    print("Topic: {} \nWords: {}".format(idx, topic ))
    print("\n")

Topic: 0 
Words: 0.024*"judol" + 0.018*"semoga" + 0.013*"situs" + 0.013*"menteri" + 0.012*"bu" + 0.010*"sukses" + 0.008*"iklan" + 0.008*"pandi" + 0.008*"guna" + 0.007*"sehat"


Topic: 1 
Words: 0.037*"transformasi" + 0.029*"tingkat" + 0.025*"masyarakat" + 0.022*"ekonomi" + 0.021*"terima" + 0.019*"teknologi" + 0.019*"perintah" + 0.013*"manfaat" + 0.012*"digitalisasi" + 0.012*"kembang"


Topic: 2 
Words: 0.024*"hukum" + 0.014*"negara" + 0.013*"terima" + 0.013*"polisi" + 0.013*"narkoba" + 0.011*"negeri" + 0.010*"institusi" + 0.010*"dia" + 0.009*"maju" + 0.008*"buat"




Dari output diatas, bisa dianalisa dari kata-kata di setiap topik bahwa:

#### Topik 0:
- **Kata-kata utama:** "judol", "semoga", "menteri", "situs", "bu"
- **Interpretasi:** Topik ini sama sekali tidak berkaitan dengan **transformasi digital**. Sebaliknya, topik ini lebih cenderung membahas **isu-isu sosial atau harapan masyarakat** terhadap figur publik seperti "menteri" dan penggunaan istilah seperti "bu". Kata-kata seperti "situs" dan "semoga" menunjukkan bahwa diskusi dalam topik ini lebih fokus pada **sosial media**, **peran pemerintah**, atau **permohonan tindakan** terhadap situasi tertentu, seperti pemblokiran situs yang dianggap merusak. Hal ini tidak terkait dengan konsep **inovasi teknologi atau digitalisasi**, melainkan lebih kepada **permintaan sosial terhadap tindakan spesifik**.

#### Topik 1:
- **Kata-kata utama:** "transformasi", "tingkat", "masyarakat", "ekonomi", "digital"
- **Interpretasi:** Topik ini sangat berkaitan dengan **transformasi digital**, terutama dalam konteks **pengembangan masyarakat dan ekonomi**. Kata-kata seperti "transformasi", "digital", dan "ekonomi" menunjukkan bahwa ini adalah diskusi tentang bagaimana **inovasi teknologi** dapat diterapkan untuk **memajukan masyarakat** dan **meningkatkan ekonomi**. Ini mencerminkan topik-topik terkait **digitalisasi** yang berperan penting dalam **mendorong pertumbuhan sosial dan ekonomi** di era modern.

#### Topik 2:
- **Kata-kata utama:** "hukum", "negara", "polisi", "narkoba", "keamanan"
- **Interpretasi:** Topik ini berfokus pada **hukum dan keamanan**, termasuk diskusi tentang **kejahatan narkoba**, **penegakan hukum**, dan **peran aparat keamanan**. Kata-kata seperti "hukum", "polisi", dan "narkoba" menunjukkan bahwa topik ini berisi percakapan tentang **isu kriminal** atau **tindakan hukum yang dilakukan pemerintah** untuk menjaga stabilitas sosial. Ini tidak terkait dengan **transformasi digital**, melainkan lebih kepada **masalah regulasi dan keamanan**.

#### Kesimpulan:
Model LDA ini membagi dokumen menjadi tiga topik dengan karakteristik sebagai berikut:
1. **Topik 0**: Tidak relevan dengan **transformasi digital**, fokus pada **isu sosial**, permintaan tindakan terhadap figur publik, atau harapan masyarakat.
2. **Topik 1**: Sangat relevan dengan **transformasi digital**, membahas **inovasi teknologi** dan dampaknya terhadap **masyarakat dan ekonomi**.
3. **Topik 2**: Berhubungan dengan **hukum dan keamanan**, fokus pada **penegakan hukum** dan **masalah sosial** seperti narkoba.

Pembagian ini memberikan gambaran yang jelas mengenai tema utama yang muncul dalam dokumen-dokumen yang dianalisis.

### Human Judgement
Dari data test

In [ ]:
X_test_lemmatized.head()

185    [sehebat, apap, teknologi, digital, kalah, ind...
78                                                    []
55     [judol, rajarela, indonesia, bu, bersihkan, ak...
137    [transformasi, digital, nasional, ina, digital...
161    [dapat, terima, isi, video, saji, pandemi, cov...
Name: cleaned_text, dtype: object

Dari output diatas, index tidak berurutan, maka perlu di reset dulu

In [ ]:
# Reset indeks DataFrame
X_test_lemmatized2 = X_test_lemmatized.reset_index(drop=True)

In [ ]:
# Menentukan topik untuk seluruh data test dan menghitung skor topik
test_shape = len(X_test_lemmatized2)
sentence, topik0, score0, topik1, score1, topik2, score2 = [], [], [], [], [], [], []

for doc in range(test_shape):
    # Menyimpan kalimat asli (tidak ditoken)
    sentence.append(" ".join(X_test_lemmatized2[doc]))  # Menggabungkan token menjadi satu string

    # Mengambil tiga topik dengan skor tertinggi
    sorted_topics = sorted(lda_model[bow_corpusTest[doc]], key=lambda tup: -1 * tup[1])

    # Inisialisasi skor untuk setiap topik
    scores = {0: "-", 1: "-", 2: "-"}

    # Menyimpan skor berdasarkan topik
    for topic, score in sorted_topics:
        scores[topic] = score

    # Menandai topik dengan relevansi tertinggi menggunakan ✓
    max_topic = max(sorted_topics, key=lambda tup: tup[1])[0]
    if max_topic == 0:
        topik0.append("✓")
        score0.append(scores[0])
        topik1.append("-")
        score1.append(scores[1])
        topik2.append("-")
        score2.append(scores[2])
    elif max_topic == 1:
        topik0.append("-")
        score0.append(scores[0])
        topik1.append("✓")
        score1.append(scores[1])
        topik2.append("-")
        score2.append(scores[2])
    elif max_topic == 2:
        topik0.append("-")
        score0.append(scores[0])
        topik1.append("-")
        score1.append(scores[1])
        topik2.append("✓")
        score2.append(scores[2])

# Menyusun hasil dalam dataframe
test_result = pd.DataFrame({
    'Kalimat': sentence,
    'Topik 0': topik0,
    'Relevansi (0)': score0,
    'Topik 1': topik1,
    'Relevansi (1)': score1,
    'Topik 2': topik2,
    'Relevansi (2)': score2
})

**Code Jugdement**

In [ ]:
test_result.iloc[10:20]

Kalimat Topik 0 Relevansi (0)  \
10  bicara tranformasi digital indonesia tranforma...       -      0.010758   
11                                  iklan judol kelia       ✓       0.77754   
12  dapat digitalisasi tingkat tolong sistem ekono...       -      0.070222   
13  tolong bu mentri hormat beserta jajar jam digi...       ✓      0.625914   
14                                        judol kabar       ✓      0.666171   
15                                              seksi       ✓      0.333333   
16  video berhasil informasi dalam transformasi di...       -      0.026322   
17  transformasi digital kunci utama tingkat mandi...       -             -   
18  semoga kembang teknologi ikuta rasa kembang di...       -      0.078387   
19  atur bu menteri anak kebawah larang gadget sos...       -      0.277002   

   Topik 1  Relevansi (1) Topik 2  Relevansi (2)  
10       ✓       0.934399       -       0.054843  
11       -       0.111288       -       0.111171  
12       ✓       0.906592       -       0.023186  
13       -       0.019770       -       0.354316  
14       -       0.167099       -       0.166730  
15       -       0.333333       -       0.333333  
16       ✓       0.537071       -       0.436607  
17       ✓       0.944165       -       0.050101  
18       -       0.304968       ✓       0.616645  
19       -       0.230484       ✓       0.492514

**Human Jugdement**

| **Kalimat**                                                                                     | **Topik 0** | **Topik 1** | **Topik 2** |
|-------------------------------------------------------------------------------------------------|-------------|-------------|-------------|
| bicara tranformasi digital indonesia tranforma...                                               | -           | ✓           | -           |
| iklan judol kelia                                                                               | ✓           | -           | -           |
| dapat digitalisasi tingkat tolong sistem ekono...                                              | -           | ✓           | -           |
| tolong bu mentri hormat beserta jajar jam digi...                                               | -           | -           | ✓           |
| judol kabar                                                                                    | ✓           | -           | -           |
| seksi                                                                                          | ✓           | -           | -           |
| video berhasil informasi dalam transformasi di...                                              | -           | ✓           | -           |
| transformasi digital kunci utama tingkat mandi...                                              | -           | ✓           | -           |
| semoga kembang teknologi ikuta rasa kembang di...                                              | -           | -           | ✓           |
| atur bu menteri anak kebawah larang gadget sos...                                              | -           | -           | ✓           |


Perbedaannya ada di kalimat 'tolong bu mentri hormat beserta jajar'. Menurut saya akan lebih cocok masuk ke topik 3, karena masih berkaitan dengan penegakan hukum ataupun masalah sosial

In [ ]:
print(df['cleaned_text'].iloc[30])

minta tolong bu mentri yg terhormat beserta jajaran di jaman digital seperti sekarang ini banyak aplikasi dan situs yg merusak generasi muda kalau di biarkan akan membahayakan saya mohon untuk semua situs dan aplikasi perusak generasi muda untuk di blokir


### Visualisasi

In [ ]:
!pip install pyLDAvis

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.6/2.6 MB 22.6 MB/s eta 0:00:00


In [ ]:
import pyLDAvis.gensim
import pickle
import pyLDAvis
import os

In [ ]:
vis = pyLDAvis.gensim.prepare(topic_model=lda_model,
                              corpus=bow_corpusTest,
                              dictionary=dictionary)
pyLDAvis.enable_notebook()
pyLDAvis.display(vis)

/usr/local/lib/python3.11/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Dari hasil visualisasi, jika kita pilih selected topic, akan muncul kata-kata yang masuk ke dalam topik yang kita pilih, dan terlihat bahwa topik 1 mengandung kata-kata yang sangat berhubungan dengan transformasi digital

## **Perbandingan**

### **Hasil Clustering:**
Hasil clustering membagi teks ke dalam dua kelompok besar berdasarkan tema utama yang dibahas:

1. **Cluster 0: Berkaitan dengan transformasi digital**
   - Kelompok ini sangat membahas tentang **transformasi digital** dan dampaknya di berbagai sektor, termasuk **pemerintahan** dan **masyarakat**. Teks dalam cluster ini sering kali membahas **digitalisasi**, **inovasi teknologi**, dan **kebijakan pemerintah** yang terkait dengan perkembangan teknologi.
   - Contohnya, ada diskusi tentang **bagaimana teknologi mengubah pola kerja dan struktur ekonomi** serta **tantangan regulasi** untuk mendukung transformasi tersebut.

2. **Cluster 1: Tidak berkaitan dengan transformasi digital**
   - Kelompok ini lebih berfokus pada **isu sosial**, seperti **motivasi pribadi**, **interaksi sosial**, dan **harapan masyarakat**. Banyak kata-kata yang terkait dengan **dukungan terhadap figur publik**, **platform media sosial**, dan **isu sosial-politik** lainnya.
   - Selain itu, cluster ini juga menyoroti topik seperti **penegakan hukum**, **keamanan**, dan **harapan terhadap perubahan di sektor sosial dan pemerintahan**, yang tidak memiliki kaitan langsung dengan transformasi digital.

### **Hasil Topic Prediction (LDA):**
LDA memecah teks menjadi tiga topik berdasarkan kata-kata yang sering muncul bersama-sama. Berikut interpretasi masing-masing topik:

1. **Topik 0: Harapan Masyarakat**  
   - Topik ini tidak berkaitan dengan **transformasi digital** dan lebih cenderung membahas isu sosial dan harapan masyarakat. Kata-kata seperti **"menteri"**, **"semoga"**, dan **"bu"** menunjukkan diskusi tentang **figur publik**, harapan terhadap perubahan kebijakan, atau ucapan dukungan terhadap tindakan sosial atau pemerintahan.

2. **Topik 1: Transformasi digital**  
   - Topik ini sangat berfokus pada **transformasi digital** dan dampaknya terhadap **masyarakat** dan **ekonomi**. Diskusi mencakup **digitalisasi**, **pengembangan teknologi**, dan **bagaimana inovasi teknologi memengaruhi kehidupan sosial** serta struktur ekonomi. Kata-kata seperti **"transformasi"**, **"masyarakat"**, dan **"ekonomi"** memperkuat tema ini.

3. **Topik 2: Hukum dan keamanan**  
   - Topik ini berkaitan dengan **masalah hukum**, **keamanan**, dan **penegakan hukum**, seperti isu-isu kriminal dan tindakan kepolisian. Kata-kata seperti **"polisi"**, **"narkoba"**, dan **"hukum"** menunjukkan fokus pada **penanganan kejahatan** serta regulasi hukum di masyarakat.

### **Perbandingan Hasil Clustering dan Topic Prediction:**
- **Cluster 0** yang berfokus pada **transformasi digital** sejalan dengan **Topik 1** dalam LDA, yang sama-sama membahas tentang **inovasi teknologi** dan dampaknya terhadap masyarakat serta ekonomi. Namun, Cluster 0 lebih menyoroti **pengaruh teknologi secara umum**, sedangkan Topik 1 memusatkan perhatian pada **transformasi sosial dan ekonomi yang disebabkan oleh teknologi**.
  
- **Cluster 1** yang tidak berkaitan dengan **transformasi digital** paling mirip dengan **Topik 0** dalam LDA, yang membahas **isu sosial**, **figur publik**, dan harapan masyarakat. Meski Cluster 1 juga menyoroti **interaksi sosial di media**, Topik 0 lebih menekankan **aspek sosial-politik**.

- **Topik 2** dalam LDA, yang berfokus pada **hukum dan keamanan**, tidak secara spesifik tergambar dalam hasil clustering. Namun, beberapa teks dalam **Cluster 1** kemungkinan berkaitan dengan tema ini, terutama yang membahas **keamanan atau penegakan hukum**.

### **Kesimpulan:**
Hasil clustering dan LDA memberikan wawasan yang saling melengkapi. **Clustering** membantu mengidentifikasi **dua tema besar**: relevansi dengan transformasi digital dan isu-isu sosial lainnya. Sementara itu, **LDA** menggali lebih dalam, membedakan topik-topik spesifik, seperti **transformasi digital**, **isu sosial-politik**, dan **hukum dan keamanan**. Kombinasi keduanya membantu memberikan gambaran yang lebih detail mengenai struktur tema dalam teks yang dianalisis.